### PR TITLE
nomis DSOS-1478: align pipelines with strategy stage 1

### PR DESCRIPTION
**Pull request by dominicrobinson-moj**
New pipeline for NOMIS AMI images.  The idea is we will have one directory per image, each with its own statefile, which will be triggered by this single pipeline.  Pipeline nicked from dso-infra-azure-fixngo which has similar structure.

Stuff to call out:
- pipeline does locking (in case multiple people working at same time) + posts PR comments.
- the scripts/terraform-* updated so they can be run with a relative path
- added a terraform-refresh.  Pipeline runs this as a separate step so refresh changes don't get posted in PR
- I've updated the pipeline version numbers to test plan/apply changes get posted into PR.
- Current terraform is move into an old-terraform/.  This will be split up as a separate PR.

### CODE REVIEW

File: .github/workflows/nomis.yml [hunk 1]
```diff
@@ -1,58 +1,366 @@
...
+        with:
+          same-branch-only: false
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Init and State Refresh
+        working-directory: "${{ env.TEAM_DIR }}/${{ matrix.project }}"
+        run: |
+          ${GITHUB_WORKSPACE}/scripts/terraform-init.sh .
+          terraform workspace select "core-shared-services-${TF_ENV}"
```
Comment (julialawrence) [inline .github/workflows/nomis.yml:227]: `terraform workspace select "core-shared-services-${TF_ENV}" || terraform workspace new "core-shared-services-${TF_ENV}"`

The workspace won't exist when a new state is created so it'll save hacking about with the workflow file every time a new pipeline's added.

**Review by julialawrence — CHANGES_REQUESTED** [review]
One suggestion in the worklow file.

File: .github/workflows/nomis.yml (hunk 1 above)
Comment (dominicrobinson-moj) [reply .github/workflows/nomis.yml:227]: Totally agree - but would like to put that in the next PR, where I will create a new terraform state, to ensure all of that is fully tested.  There maybe other tweaks required as well

**Review by julialawrence — DISMISSED** [review]

**Review by dms1981 — APPROVED** [review]

**Review by SteveLinden — APPROVED** [review]

**Review by dms1981 — COMMENTED** [review]
Given that `terraform refresh` is a deprecated command, have you considered an alternative way to achieve a refresh - such as a flag for `terraform apply`?
https://www.terraform.io/cli/commands/refresh

File: scripts/terraform-apply.sh [hunk 2]
```diff
@@ -11,16 +11,14 @@ set -e
 # are exposed via terraform apply.
 
 # Make redact-output.sh executable
-chmod +x ./scripts/redact-output.sh
+chmod +x $(dirname $0)/redact-output.sh
```
Comment (dms1981) [inline scripts/terraform-apply.sh:14]: Given that this script will always run from this directory, is there any benefit from using `dirname` to derive the directory name? Does an absolute path to the `redact-output.sh` offer any benefits?

Comment (dominicrobinson-moj) [reply scripts/terraform-apply.sh:14]: In this pipeline, I'm calling this from the terraform project instead, like this for example.  So I don't need to be forever changing directory, e.g. with the workspace select.

      - name: Init and State Refresh
        working-directory: "${{ env.TEAM_DIR }}/${{ matrix.project }}"
        run: |
          ${GITHUB_WORKSPACE}/scripts/terraform-init.sh .
          terraform workspace select "core-shared-services-${TF_ENV}"
          ${GITHUB_WORKSPACE}/scripts/terraform-refresh.sh .

**Review by ewastempel — APPROVED** [review]
Nice! a bit heavy, so hopefully did not miss anything :)

File: scripts/terraform-apply.sh [hunk 3]
```diff
@@ -11,16 +11,14 @@ set -e
...
   exit 1
 fi
 
-if [ ! -z "$2" ]; then
-  options="$2"
-  terraform -chdir="$1" apply -input=false -no-color -auto-approve $options | ./scripts/redact-output.sh
-else
-  terraform -chdir="$1" apply -input=false -no-color -auto-approve | ./scripts/redact-output.sh  
-fi
+terraform -chdir="$project_dir" apply -input=false -no-color -auto-approve $@ | $(dirname $0)/redact-output.sh
```
[the remark this answers is not in the document]
Comment (dominicrobinson-moj) [reply scripts/terraform-apply.sh:24]: > We are losing the ability to pass in Terraform options here

It does still work, in fact previously you had to quote all the extra arguments as the 2nd arg which for me wasn't intuitive, or at least, I fell foul of it.  So now all additional arguments are passed into terraform via the $@ at the end of line 24.  So I call like this for example `terraform-plan.sh . -refresh=false -detailed-exitcode`

File: scripts/terraform-plan.sh [hunk 4]
```diff
@@ -11,16 +11,14 @@ set -e
...
   exit 1
 fi
 
-if [ ! -z "$2" ]; then
-  options="$2"
-  terraform -chdir="$1" plan -input=false -no-color $options | ./scripts/redact-output.sh
-else
-  terraform -chdir="$1" plan -input=false -no-color | ./scripts/redact-output.sh
-fi
+terraform -chdir="$project_dir" plan -input=false -no-color $@ | $(dirname $0)/redact-output.sh
```
[the remark this answers is not in the document]
Comment (dominicrobinson-moj) [reply scripts/terraform-plan.sh:24]: > As above we are losing the ability to pass in options

It does still work, in fact previously you had to quote all the extra arguments as the 2nd arg which for me wasn't intuitive, or at least, I fell foul of it. So now all additional arguments are passed into terraform via the $@ at the end of line 24. So I call like this for example terraform-plan.sh . -refresh=false -detailed-exitcode

**Review by davidkelliott — APPROVED** [review]
Approving this PR, public actions are to be addressed in a separate PR next sprint

**Review by wullub — APPROVED** [review]

**Pull request merged**